src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+          }
+      }
+      else {
+        (-1, Integer.MAX_VALUE, Integer.MAX_VALUE)
+      }
+      // Make sure we are within the parameter limits and update barcode metrics if necessary
+      if (maxMismatches < bestMismatches || maxNoCalls < numNoCalls || (secondBestMismatches - bestMismatches) < minMismatchDelta) {
+        noMatchBarcodeMetric.reads +=  1
+        noMatchBarcodeMetric.pf_reads += 1
+        None
+      }
+      else {
+        val sampleBarcode = sampleSheet.get(bestSampleOrdinal - 1).sampleBarcode.getOrElse(unreachable(s"Sample barcode required for sample"))
+        val bestBarcodeMetric: SampleBarcodeMetric = this.barcodeToMetrics.getOrElse(sampleBarcode.concatenatedBarcode, unreachable("No metric for sample"))
+        bestBarcodeMetric.reads += 1
+        bestBarcodeMetric.pf_reads += 1
+        if (bestMismatches == 0) {
+          bestBarcodeMetric.perfect_matches += 1
+          bestBarcodeMetric.pf_perfect_matches += 1
+        }
+        else if (bestMismatches == 1) {
+          bestBarcodeMetric.one_mismatch_matches += 1
+          bestBarcodeMetric.pf_one_mismatch_matches += 1
+        }
+        Some(bestSampleOrdinal)
+      }
+    }
+
+    /** Creates a SAM record (always paired) */
+    private def makeSAMRecord(header: SAMFileHeader, fastqRecord: FastqRecord, readStructure: ReadStructure,
+                              readGroupId: String, pairedEnd: Boolean, firstOfPair: Boolean): SAMRecord = {
+      val (readbases, readQualities) = readStructure.structureReadWithQualities(
+        bases     = fastqRecord.bases,
+        qualities = fastqRecord.quals,
+        strict=false
+      ).map { case SubRead(b, q, s) => (b, q.getOrElse(unreachable())) }.unzip
+      val record: SAMRecord = new SAMRecord(header)
+      val readBases = readbases.mkString.getBytes
+      val baseQualities = readQualities.mkString.getBytes
+
+      record.setReadName(fastqRecord.name)
+      record.setReadBases(readBases)
+      record.setReadUnmappedFlag(true)
+      if (pairedEnd) {
+        record.setReadPairedFlag(true)
+        record.setMateUnmappedFlag(true)
+        if (firstOfPair) record.setFirstOfPairFlag(true)
+        else record.setSecondOfPairFlag(true)
+      }
+      record.setAttribute(ReservedTagConstants.READ_GROUP_ID, readGroupId)
+
+      convertQuality(baseQualities, qualityFormat)
+      baseQualities.foreach {
+        qual =>
+          val uQual: Int = qual & 0xff
+          if (uQual < minQ || uQual > maxQ) {
+            throw new IllegalStateException(s"Base quality $uQual is not in the range $minQ ... $maxQ for read ${fastqRecord.header}")
+          }
+      }
+      record.setBaseQualities(baseQualities)
+
+      record
+    }
+
+    /** Based on the type of quality scores coming in, converts them to a numeric byte[] in phred scale. */
+    private[fastq] def convertQuality(quals: Array[Byte], version: FastqQualityFormat) {
```

Comment:
I think I've figure out that it's actually `Unit` and all the other methods modify the byte[] in place?